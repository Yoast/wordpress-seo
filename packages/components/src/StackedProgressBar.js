import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const StackedProgressBarContainer = styled.div`
	margin: 8px 0;
	height: ${ props => props.barHeight };
	overflow: hidden;
`;

const StackedProgressBarProgress = styled.span`
	display: inline-block;
	vertical-align: top;
	width: ${ props => `${ props.progressWidth }%` };
	background-color: ${ props => props.progressColor };
	height: 100%;
`;

StackedProgressBarProgress.propTypes = {
	progressWidth: PropTypes.number.isRequired,
	progressColor: PropTypes.string.isRequired,
};

/**
 * The StackedProgressBar component.
 *
 * @param {object} props The component's props.
 *
 * @returns {ReactElement} The StackedProgressBarContainer component.
 */
const StackedProgressBar = ( props ) => {
	let totalValue = 0;
	for ( let i = 0; i < props.items.length; i++ ) {
		props.items[ i ].value = Math.max( props.items[ i ].value, 0 );
		totalValue += props.items[ i ].value;
	}

	if ( totalValue <= 0 ) {
		return null;
	}

	return (
		<StackedProgressBarContainer
			className={ props.className }
			barHeight={ props.barHeight }
		>
			{ props.items.map( ( item, index ) =>
				<StackedProgressBarProgress
					className={ `${ props.className }__part` }
					key={ index }
					progressColor={ item.color }
					progressWidth={ item.value / totalValue * 100 }
				/>
			) }
		</StackedProgressBarContainer>
	);
};

StackedProgressBar.propTypes = {
	className: PropTypes.string,
	items: PropTypes.arrayOf(
		PropTypes.shape( {
			value: PropTypes.number.isRequired,
			color: PropTypes.string.isRequired,
		} )
	),
	barHeight: PropTypes.string,
};

StackedProgressBar.defaultProps = {
	className: "stacked-progress-bar",
	barHeight: "24px",
};

export default StackedProgressBar;
