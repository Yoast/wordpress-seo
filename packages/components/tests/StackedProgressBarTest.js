import React from "react";
import renderer from "react-test-renderer";

import StackedProgressBar from "../src/StackedProgressBar";

test( "the StackedProgressBar matches the snapshot", () => {
	const items = [
		{
			value: 7,
			color: "#F00",
		},
		{
			value: 216,
			color: "#FF0",
		},
		{
			value: 1134,
			color: "#0FF",
		},
		{
			value: 76,
			color: "#333",
		},
	];

	const component = renderer.create(
		<StackedProgressBar
			items={ items }
		/>
	);

	const tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );
