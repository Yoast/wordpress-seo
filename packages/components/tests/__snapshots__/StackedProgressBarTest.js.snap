// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the StackedProgressBar matches the snapshot 1`] = `
.c0 {
  margin: 8px 0;
  height: 24px;
  overflow: hidden;
}

.c1 {
  display: inline-block;
  vertical-align: top;
  width: 0.48848569434752265%;
  background-color: #F00;
  height: 100%;
}

.c2 {
  display: inline-block;
  vertical-align: top;
  width: 15.073272854152128%;
  background-color: #FF0;
  height: 100%;
}

.c3 {
  display: inline-block;
  vertical-align: top;
  width: 79.13468248429868%;
  background-color: #0FF;
  height: 100%;
}

.c4 {
  display: inline-block;
  vertical-align: top;
  width: 5.303558967201675%;
  background-color: #333;
  height: 100%;
}

<div
  className="c0 stacked-progress-bar"
>
  <span
    className="c1 stacked-progress-bar__part"
  />
  <span
    className="c2 stacked-progress-bar__part"
  />
  <span
    className="c3 stacked-progress-bar__part"
  />
  <span
    className="c4 stacked-progress-bar__part"
  />
</div>
`;
